File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        pLen = len(prices)
        globalMin = prices[0]
        maxProfit = 0
        
        i = 0
        while i < pLen:
            
            price = prices[i]
            
            if price < globalMin:
                globalMin = price
            
            if i == 0 or (i > 0 and price < prices[i - 1]):
                localMin = price 
                i += 1
                
                while i < pLen and prices[i] >= prices[i - 1]:
                    i += 1
                
                maxProfit = max(prices[i - 1] - globalMin, maxProfit + prices[i - 1] - localMin)
            else:
                i += 1

            
        return maxProfit
               

File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<int> dp(prices.size(),0);
        int curmin = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            dp[i] = max(prices[i] - prices[i - 1] + dp[i - 1], prices[i] - curmin);
            dp[i] = max(dp[i], dp[i - 1]);
            curmin = min(curmin, prices[i]);
        }
        return dp.back();
    }
};